### PR TITLE
Added NativeElement

### PR DESCRIPTION
**Pull request by Tom01098**
I've created and added a NativeElement container - it's just a copy of [my repo](https://github.com/Tom01098/NativeElement-UnityCollections). The intention is to allow a single blittable element to be passed into a job.

### CODE REVIEW

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+
+/// <summary>
+/// NativeElement is a wrapper for a single blittable element to be passed between managed and unmanaged code in jobs
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:8]: Line is longer than 80 characters.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+/// </summary>
+/// <typeparam name="T">The blittable type to be stored</typeparam>
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:10]: XML elements should be separated like this:

```csharp
/// <summary>
/// The summary
/// </summary>
/// 
/// <typeparam name="T">
/// Explain the type param
/// </typeparam>
```

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+[StructLayout(LayoutKind.Sequential)]
+[NativeContainer]
+[DebuggerDisplay("{Element}")]
+public unsafe struct NativeElement<T> : IDisposable where T : unmanaged
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:14]: `where` goes on the next line, indented.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+{
+    // The pointer to the element
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:16]: Describe with an xml-doc style comment.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    [NativeDisableUnsafePtrRestriction]
+    T* m_ptr;
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:18]: Fields should be named like `m_MyVar` for consistency with Unity-mandated field names like `m_MinIndex`.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+    // Safety
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:21]: Describe with an xml-doc style comment.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    AtomicSafetyHandle m_Safety;
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:22]: Should be explicitly marked `private`.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+    [NativeSetClassTypeToNullOnSchedule]
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:24]: Describe with an xml-doc style comment.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    DisposeSentinel m_DisposeSentinel;
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:25]: Should be explicitly marked `private`.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+#endif
+
+    // The allocator label
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:28]: Describe with an xml-doc style comment.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    Allocator m_AllocatorLabel;
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:29]: Should be explicitly marked `private`.

Comment (jacksondunstan) [reply JacksonDunstanNativeCollections/NativeElement.cs:29]: Rename to `m_Allocator` for consistency with other types.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+    // Constructor which only takes a label and then calls the main constructor with an instance of T
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:31]: Describe with an xml-doc style comment.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    public NativeElement(Allocator label) : this(new T(), label) { }
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:32]: Remove in favor of a default parameter to the other constructor: `T element = default(T)`.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+
+    // Main constructor logic that takes an element and a label
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:34]: Describe with an xml-doc style comment.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    public NativeElement(T element, Allocator label)
+    {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+        // Label check
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:38]: Reword to "Require a valid allocator" as in other types.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    // The allocator label
+    Allocator m_AllocatorLabel;
+
+    // Constructor which only takes a label and then calls the main constructor with an instance of T
+    public NativeElement(Allocator label) : this(new T(), label) { }
+
+    // Main constructor logic that takes an element and a label
+    public NativeElement(T element, Allocator label)
+    {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:37]: Remove `#if`...`#endif` wrapping as this check should occur all the time.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        // Label check
+        if (label <= Allocator.None)
+            throw new ArgumentException("NativeElement must be allocated using Job, TempJob or Persistent");
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:40]: Add curly braces around this line.

Comment (jacksondunstan) [reply JacksonDunstanNativeCollections/NativeElement.cs:40]: Change to match other types in the project:
```csharp
throw new ArgumentException(
	"Allocator must be Temp, TempJob or Persistent",
	"allocator");
```

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+#endif
+
+        // Label set
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:43]: Reword to match other types:
```csharp
// Store the allocator for future allocation and deallocation
// operations
```

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+        m_AllocatorLabel = label;
+
+        // Allocate memory for a single T
+        m_ptr = (T*)UnsafeUtility.Malloc(UnsafeUtility.SizeOf<T>(), 1, label);
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+        // Create DisposeSentinel and AtomicSafetyHandle
+        DisposeSentinel.Create(out m_Safety, out m_DisposeSentinel, 0, label);
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:51]: Support both Unity 2018.2 and Unity 2018.3 versions of this function as other types do:
```csharp
#if UNITY_2018_3_OR_NEWER
	DisposeSentinel.Create(
		out m_Safety,
		out m_DisposeSentinel,
		0,
		allocator);
#else
	DisposeSentinel.Create(out m_Safety, out m_DisposeSentinel, 0);
#endif
```

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+#endif
+
+        // Create element to avoid unitialised data
+        Element = element;
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:55]: Remove property function call and safety check by replacing with `m_Ptr = element`.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+
+        // Allocate memory for a single T
+        m_ptr = (T*)UnsafeUtility.Malloc(UnsafeUtility.SizeOf<T>(), 1, label);
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+        // Create DisposeSentinel and AtomicSafetyHandle
+        DisposeSentinel.Create(out m_Safety, out m_DisposeSentinel, 0, label);
+#endif
+
+        // Create element to avoid unitialised data
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:54]: Reword for improved accuracy. For example, `Set the initial value.`

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+        Element = element;
+    }
+
+    // Property for the Element stored
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:58]: Describe with an xml-doc style comment.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+    public T Element
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:59]: Rename to `Value` for consistency with `NativeIntPtr`.

File: JacksonDunstanNativeCollections/NativeElement.cs (hunk 5 above)
Comment (jacksondunstan) [reply JacksonDunstanNativeCollections/NativeElement.cs:18]: Rename to `m_Buffer` for consistency with `NativeIntPtr` and to support `[NativeContainerSupportsDeallocateOnJobCompletion]`.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+using Unity.Collections;
+using Unity.Collections.LowLevel.Unsafe;
+
+/// <summary>
+/// NativeElement is a wrapper for a single blittable element to be passed between managed and unmanaged code in jobs
+/// </summary>
+/// <typeparam name="T">The blittable type to be stored</typeparam>
+[StructLayout(LayoutKind.Sequential)]
+[NativeContainer]
+[DebuggerDisplay("{Element}")]
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:13]: Add `[NativeContainerSupportsDeallocateOnJobCompletion]` for consistency with `NativeIntPtr` and to allow this feature.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+        m_ptr = (T*)UnsafeUtility.Malloc(UnsafeUtility.SizeOf<T>(), 1, label);
+
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+        // Create DisposeSentinel and AtomicSafetyHandle
+        DisposeSentinel.Create(out m_Safety, out m_DisposeSentinel, 0, label);
+#endif
+
+        // Create element to avoid unitialised data
+        Element = element;
+    }
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:56]: Add a `RequireBlittable` call like in other types.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+
+    // Property for the Element stored
+    public T Element
+    {
+        get
+        {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:63]: Move this check to a `RequireReadAccess` function.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+            // Read check
+            AtomicSafetyHandle.CheckReadAndThrow(m_Safety);
+#endif
+
+            // Return the element
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:68]: Remove or reword comment as it currently describes what is obvious.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+            return *m_ptr;
+        }
+        set
+        {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            // Write check
+            AtomicSafetyHandle.CheckWriteAndThrow(m_Safety);
+#endif
+
+            // Set the element
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:78]: Remove or reword comment as it currently describes what is obvious.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+            // Read check
+            AtomicSafetyHandle.CheckReadAndThrow(m_Safety);
+#endif
+
+            // Return the element
+            return *m_ptr;
+        }
+        set
+        {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:73]: Move this check to a `RequireWriteAccess` function.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+            // Write check
+            AtomicSafetyHandle.CheckWriteAndThrow(m_Safety);
+#endif
+
+            // Set the element
+            *m_ptr = value;
+        }
+    }
+
+    // Dispose of all resources
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:83]: Replace with a copy of `NativeIntPtr.Dispose` to improve commenting, support for both 2018.2 and 2018.3, add `[WriteAccessRequired]`, add `RequireWriteAccess`, and set `m_Buffer` to `null`.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+        {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+            // Read check
+            AtomicSafetyHandle.CheckReadAndThrow(m_Safety);
+#endif
+
+            // Return the element
+            return *m_ptr;
+        }
+        set
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:71]: Add `[WriteAccessRequired]`.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+    // Dispose of all resources
+    public void Dispose()
+    {
+#if ENABLE_UNITY_COLLECTIONS_CHECKS
+        DisposeSentinel.Dispose(ref m_Safety, ref m_DisposeSentinel);
+#endif
+
+        UnsafeUtility.Free(m_ptr, m_AllocatorLabel);
+    }
+}
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:92]: Add `IsCreated` property.

File: JacksonDunstanNativeCollections/NativeElement.cs [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+#endif
+
+    // The allocator label
+    Allocator m_AllocatorLabel;
+
+    // Constructor which only takes a label and then calls the main constructor with an instance of T
+    public NativeElement(Allocator label) : this(new T(), label) { }
+
+    // Main constructor logic that takes an element and a label
+    public NativeElement(T element, Allocator label)
```
Comment (jacksondunstan) [inline JacksonDunstanNativeCollections/NativeElement.cs:35]: Rename `element` to `initialValue` for consistency with other types, move it to the end of the parameter list, and add a default value: `T initialValue = default(T)`.

Comment (jacksondunstan) [reply JacksonDunstanNativeCollections/NativeElement.cs:35]: Rename `label` to `allocator` for consistency with other types.

**Review by jacksondunstan — CHANGES_REQUESTED** [review]
Hi @Tom01098,

This looks like a nice addition to NativeCollections! Before I merge though, I'd like to see a few changes made so that it better fits in with the other types (e.g. `NativeLinkedList`) so I've pointed out a number of issues in the file. Most of them should be trivial to change as they mostly relate to consistent styling, but in a few cases they do relate to functionality. One piece that's completely missing though is unit tests. Every other type is thoroughly tested and the same should be true here.

Thanks again for the submission,
-Jackson

**Comment by Tom01098** [comment]
Hey, thanks for the detailed review :) I'll work on it - not familiar with NUnit though so I might be forever ;)

File: JacksonDunstanNativeCollections/NativeElement.cs (hunk 25 above)
Comment (Tom01098) [reply JacksonDunstanNativeCollections/NativeElement.cs:56]: Hi, I've made all the other changes you requested, but this one is a bit confusing to me - the type parameter T is already constrained to be unmanaged - would you like me to change it to be constrained to a struct and then have a blittable check?

Comment (jacksondunstan) [reply JacksonDunstanNativeCollections/NativeElement.cs:56]: Yes, that's necessary for compatibility with Unity 2018.2 when C# 7 wasn't yet supported and for consistency with Unity's types such as `NativeArray<T>`.

**Comment by jacksondunstan** [comment]
> Hey, thanks for the detailed review :) I'll work on it - not familiar with NUnit though so I might be forever ;)

Hi @Tom01098, please take your time! :) NUnit is well worth learning as other xUnit frameworks are common in other languages such as Java and the skills learned in one tend to port well to the others.

I missed one item in my initial review: the type should be renamed to `NativePtr` for consistency with `NativeIntPtr`, `NativePerJobThreadIntPtr`, `NativeLongPtr`, and `NativePerJobThreadLongPtr` and also because it's not really an element of anything like an array but rather a pointer to a single value.

**Comment by torsina** [comment]
@Tom01098 update ?

**Comment by Tom01098** [comment]
> @Tom01098 update ?

Sorry about this; I no longer work with Unity and so should have closed this PR a while ago. I assume the basic concept of what I did here still works in newer versions of the Job System.

**Pull request closed without merging**